airbyte-integrations/connectors/destination-s3/src/main/java/io/airbyte/integrations/destination/s3/csv/S3CsvOutputFormatter.java

```diff
@@ -0,0 +1,269 @@
+  private final S3CsvFormatConfig formatConfig;
+  private final AmazonS3 s3Client;
+  private final AirbyteStream stream;
+  private final DestinationSyncMode syncMode;
+  private final List<String> sortedHeaders;
+  private final String outputPrefix;
+  private final StreamTransferManager uploadManager;
+  private final MultiPartOutputStream outputStream;
+  private final CSVPrinter csvPrinter;
+
+  public S3CsvOutputFormatter(S3DestinationConfig config,
+                              AmazonS3 s3Client,
+                              ConfiguredAirbyteStream configuredStream,
+                              Timestamp uploadTimestamp)
+      throws IOException {
+    this.config = config;
+    this.formatConfig = (S3CsvFormatConfig) config.getFormatConfig();
+    this.s3Client = s3Client;
+    this.stream = configuredStream.getStream();
+    this.syncMode = configuredStream.getDestinationSyncMode();
+    this.sortedHeaders = getSortedFields(configuredStream.getStream().getJsonSchema(),
+        formatConfig);
+
+    // prefix: <bucket-path>/<source-namespace-if-exists>/<stream-name>
+    // filename: <upload-date>-<upload-millis>.csv
+    // full path: <bucket-name>/<prefix>/<filename>
+    this.outputPrefix = getOutputPrefix(config.getBucketPath(), stream);
+    String outputFilename = getOutputFilename(uploadTimestamp);
+    String objectKey = String.join("/", outputPrefix, outputFilename);
+
+    LOGGER.info("Full S3 path for stream '{}': {}/{}", stream.getName(), config.getBucketName(),
+        objectKey);
+
+    // The stream transfer manager lets us greedily stream into S3. The native AWS SDK does not
+    // have support for streaming multipart uploads. The alternative is first writing the entire
+    // output to disk before loading into S3. This is not feasible with large input.
+    // Data is chunked into parts during the upload. A part is sent off to a queue to be uploaded
+    // once it has reached it's configured part size.
+    // Memory consumption is queue capacity * part size = 1 * 5 = 5 MB at current configurations.
+    this.uploadManager = new StreamTransferManager(config.getBucketName(), objectKey,
+        s3Client)
+            .numUploadThreads(1)
+            .queueCapacity(1)
+            .partSize(S3DestinationConstants.DEFAULT_PART_SIZE_MD);
+    // We only need one output stream as we only have one input stream. This is reasonably performant.
+    this.outputStream = uploadManager.getMultiPartOutputStreams().get(0);
+    this.csvPrinter = new CSVPrinter(new PrintWriter(outputStream, true, StandardCharsets.UTF_8),
+        CSVFormat.DEFAULT.withQuoteMode(QuoteMode.ALL)
+            .withHeader(getHeaders(sortedHeaders).toArray(new String[0])));
+  }
+
+  /**
+   * Get a sorted field list in the json object so that this object can be iterated through with a
+   * defined order later on.
+   */
+  static List<String> getSortedFields(JsonNode jsonSchema, S3CsvFormatConfig formatConfig) {
```

Comment:
this method is a little misleading, it says it's sorting but it's actually also filtering/choosing headers. Can we move the decision of whether to flatten or un-flatten (and the resulting headers) outside of this method (e.g: in getHeaders) and have the sorting be a straightforward sort of whatever inputs are provided?

Reply:
Good point. I extracted the logics to a set of dedicated classes.